src/components/flowchart-wrapper/flowchart-wrapper.js

```diff
@@ -35,63 +35,158 @@ const linkToFlowchartInitialVal = {
+
+      if (hasModularPipeline) {
+        onToggleModularPipelineExpanded(modularPipeline);
+      }
+      onToggleNodeSelected(nodeId);
+      resetErrorMessage();
+    } else {
+      setErrorMessage(errorMessages.node);
+      setInvalidUrl(true);
+    }
+  };
+
+  const redirectToSelectedPipeline = () => {
+    const pipelineId = searchParams.get(params.pipeline);
+    const foundPipeline = pipelines.find((id) => id === pipelineId);
+
+    if (!foundPipeline) {
+      setErrorMessage(errorMessages.pipeline);
+      setInvalidUrl(true);
+    }
+  };
+
+  const redirectToFocusedNode = () => {
+    const focusedId = searchParams.get(params.focused);
+    const foundModularPipeline = modularPipelinesTree[focusedId];
+
+    if (foundModularPipeline) {
+      onToggleModularPipelineActive(focusedId, true);
+      onToggleFocusMode(foundModularPipeline.data);
+      resetErrorMessage();
+    } else {
+      setErrorMessage(errorMessages.modularPipeline);
+      setInvalidUrl(true);
+    }
+  };
+
   useEffect(() => {
-    setReload(true);
+    window.addEventListener('popstate', () => setUsedNavigationBtn(true));
 
+    return () => {
+      window.removeEventListener('popstate', () => setUsedNavigationBtn(false));
+    };
+  }, []);
+
+  useEffect(() => {
     const linkToFlowchart = loadLocalStorage(localStorageFlowchartLink);
     setGoBackToExperimentTracking(linkToFlowchart);
   }, []);
 
+  /**
+   * To handle redirecting to different location via URL, eg: selectedNode, focusNode, etc
+   * We only need to call these matchPath actions either when:
+   * graphRef.current is null means the page first loads
+   * or when user navigating through the back and forward button
+   * or when the invalidUrl is true, that means the user enters some thing wrong in the URL
+   * and we should allow them to reset it by clicking on a different node on the list
+   */
   useEffect(() => {
-    const timer = setTimeout(() => {
-      setReload(false);
-    }, 200);
+    const isGraphEmpty = Object.keys(graph).length === 0;
 
-    return () => clearTimeout(timer);
-  }, []);
+    if (
+      (graphRef.current === null || usedNavigationBtn || invalidUrl) &&
+      !isGraphEmpty
+    ) {
+      if (matchedFlowchartMainPage) {
+        onToggleNodeSelected(null);
+        onToggleFocusMode(null);
 
-  const { errorMessage, invalidUrl } = useRedirectLocationInFlowchart(
-    flags,
-    fullNodeNames,
-    modularPipelinesTree,
-    nodes,
-    onLoadNodeData,
-    onToggleFocusMode,
-    onToggleModularPipelineActive,
-    onToggleModularPipelineExpanded,
-    onUpdateActivePipeline,
-    pipelines,
-    reload
-  );
+        resetErrorMessage();
+      }
+
+      if (matchedSelectedNodeName || matchedSelectedNodeId) {
+        redirectToSelectedNode();
+      }
+
+      if (matchedSelectedPipeline) {
+        redirectToSelectedPipeline();
+      }
+
+      if (matchedFocusedNode) {
+        redirectToFocusedNode();
+      }
```

Comment:
[Nit] Do we follow any heirarchy in ordering the if checks for readability ? 
FlowchartMainPage -> selectedPipeline -> selectedNode -> focusedNode ?

Reply:
hmm i don't think we follow that rule inside the component. We need it for CSS as all the lines are next to each other so it helps with readability.